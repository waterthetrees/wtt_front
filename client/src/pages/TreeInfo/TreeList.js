import React from 'react';
import { Card } from '../../components/Cards/Card';
import { NavLink } from 'react-router-dom';
import './TreeInfo.scss';
const routeName = 'treeinfo';

export default function TreeList({ data }) {
  return (
    <div className="treeinfolist">
      <div className="treeinfolist__header">
        <p>
          We encourage you to plant medium and larger-sized trees, as they
          provide greater benefits to the city than smaller trees. Young trees
          require 1.5 inches of rain or 25 gallons of water per week for the
          first three years to establish their roots.
        </p>
      </div>

      {data.map((tree, index) => {
        const nameFormatted = tree?.common.toLowerCase().replace(/ /g, '-');
        const urlRoute = `/${routeName}/${nameFormatted}`;
        const { notes, ...treeRest } = tree;
        return (
          <NavLink
            className="treeinfolist__link"
            to={{
              pathname: urlRoute,
            }}
            state={tree}
            key={`${tree?.scientific}-${index}`}
          >
            <Card {...treeRest} />
          </NavLink>
        );
      })}
    </div>
  );
}

{
  /* <div className="treeinfo__list">
  {dataList &&
    dataList.map((tree, index) => {
      const nameFormatted = tree?.common.toLowerCase().replace(/ /g, '-');
      const urlRoute = `/${routeName}/${nameFormatted}`;
      const { notes, ...treeRest } = tree;
      return (
        <NavLink
          className="treeinfo__link"
          to={{
            pathname: urlRoute,
          }}
          state={tree}
          key={`${tree?.scientific}-${index}`}
        >
          <Card {...treeRest} />
        </NavLink>
      );
    })}
</div>; */
}
