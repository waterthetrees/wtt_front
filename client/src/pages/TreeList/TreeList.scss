$border-color: #ff4040;
$checkbox-size: 1rem;
$checkedbox-size: 1.2rem;
$black: black;
$silver: silver;
$white-mix: mix($silver, white, 20%);
$white: white;
$backgroundColor: white;

$borderRadius: 4px;
.treecare {
  font-family: arial, sans-serif;
  font-size: medium;
  color: $black;
  font-weight: 700;
  text-transform: uppercase;
  margin: 5px;
  cursor: pointer;
}

.treelist {
  $self: &;
  background-color: $silver;
  display: flex;
  flex-direction: column;
  justify-content: right;
  height: 100%;

  &__header {
    font-size: large;
    font-weight: 700;
    margin-bottom: 20px;
  }

  &__content {
    padding: 200px 10px;
  }

  &__content-small {
    padding: 100px 10px;
  }

  @media screen and (max-width: 768px) {
    &__content-small {
      padding: 135px 10px;
    }
  }

  @media screen and (max-width: 480px) {
    &__content {
      padding: 310px 5px;
    }
  }
}

.treelistheader {
  $self: &;
  background-image: url(../../assets/images/treelist/treelist2.png);
  background-repeat: no-repeat;
  background-attachment: fixed;
  background-size: 100%;
  display: flex;
  flex-direction: row;
  height: max-content;
  color: black;
  background-color: gray;
  padding: 58px 10px 0;
  gap: 10px;
  margin-bottom: 1px;
  justify-content: space-between;

  a {
    color: darkgray;
    text-decoration: none;
    font-size: medium;
    font-weight: 300;
  }

  &__header {
    width: 150px;
    font-size: medium;
    padding: 1rem;
  }

  &__section {
    background-color: white;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: stretch;
    gap: 10px;
    border-radius: $borderRadius;
    margin-bottom: 2px;
  }
  &__section:first-child {
    background-color: transparent;
  }
  &__section-data-select {
    background-color: $white;
    font-size: large;
    min-width: 286px;
  }

  &__section-title {
    font-size: medium;
    font-weight: 600;
    padding: 2px 5px;
  }
  &__section-item-toggle {
    background-color: transparent;
    max-width: fit-content;
    border-radius: $borderRadius;
  }

  &__section-item {
    display: flex;
    flex-direction: column;
  }

  &__section-item:focus-visible,
  &__section-item:hover {
    outline: none;
  }

  &__section-item-label {
    font-weight: 400;
    font-size: medium;
  }

  @media screen and (max-width: 1024px) {
    background-size: 100%;
  }
  @media screen and (max-width: 768px) {
    background-size: 100vh;
  }
  @media screen and (max-width: 600px) {
    background-size: 100vh;
    &__section {
      gap: 5px;
    }
    &__section-item {
      flex-direction: row;
      justify-content: space-between;
      gap: 5px;
    }
  }

  @media screen and (max-width: 480px) {
    background-size: auto 100%;

    flex-direction: column;
    &__header {
      width: 100%;
    }
    &__section-item {
      flex-direction: row;
      justify-content: space-between;
    }
  }
}

.treelistfilters {
  display: inline-flex;
  flex-direction: row;
  padding: 0;
  border-radius: $borderRadius;
  background-color: white;
  height: fit-content;
  padding: 5px;
  margin-bottom: 10px;

  &__filter {
    display: flex;
    flex-direction: column;
    padding: 8px;
  }
  &__filter-item {
    display: flex;
    flex-direction: column;
  }
  &__filter-item-checkbox {
    color: green;
    width: 20px;
    height: 20px;
    vertical-align: middle;
    position: relative;
    cursor: pointer;
    margin: 0 3px;

    input {
      position: absolute;
      opacity: 0;
      cursor: pointer;
      height: 0;
      width: 0;
      border-radius: 0;
    }

    input:hover {
      background-color: #2196f3;
    }

    /* When the checkbox is checked, add a blue background */
    input:checked {
      background-color: #2196f3;
    }
    input[type='checkbox' i] {
      border-radius: 0;
    }
  }

  &__filter-item-checkbox:active,
  &__filter-item-checkbox:checked:active {
    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05),
      inset 0px 1px 3px rgba(0, 0, 0, 0.1);
  }

  &__filter-item-checkbox:checked {
    background-color: #e9ecee;
    border: 1px solid #adb8c0;
    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05),
      inset 0px -15px 10px -12px rgba(0, 0, 0, 0.05),
      inset 15px 10px -12px rgba(255, 255, 255, 0.1);
    color: #99a1a7;
  }

  &__filter-item-label {
    font-weight: 400;
    font-size: medium;
  }

  @media screen and (max-width: 600px) {
    &__filter {
      padding: 0 2px;
    }
    &__filter-item-checkbox {
      // margin: 0 2px 0 0;
      margin: 0 30px 0 50px;
      display: inline-block;
    }
    &__filter-item-label {
      font-weight: 400;
      font-size: small;
    }
  }

  @media screen and (max-width: 480px) {
    flex-direction: column;
    font-size: medium;

    &__header {
      width: 100%;
    }

    &__filter {
      display: inline-flex;
      flex-direction: row;
      padding: 0;
    }
    &__filter-item {
      flex-direction: row;
      justify-content: space-between;
    }
    &__filter-item-checkbox {
      margin: 0 300px 0 500px;
      display: inline-block;
    }
  }
}

.treelistcards {
  $self: &;
  align-self: center;

  &__container {
    column-count: 5;
    column-gap: 1.25rem;
    width: 100%;
    max-width: 1200px;
  }

  &__header {
    font-size: large;
  }

  &__link {
    break-inside: avoid;
  }

  &__link:focus,
  &__link:hover,
  &__link:visited,
  &__link:link,
  &__link:active {
    text-decoration: none;
    opacity: 0.8;
  }

  &__image {
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: bottom;
  }

  &__image img {
    border-radius: 4px 4px 0 0;
  }

  &__item {
    color: gray;
    padding: 0 0.5rem;
  }

  &__item:first-letter {
    font-size: 1em;
    text-transform: uppercase;
  }

  &__item-notes {
    padding: 0.5rem;
    max-width: 250px;
    font-weight: 500;
  }

  @media screen and (max-width: 1024px) {
    &__container {
      column-count: 4;
    }
  }
  @media screen and (max-width: 768px) {
    &__container {
      column-count: 3;
    }
  }
  @media screen and (max-width: 600px) {
    &__container {
      column-count: 2;
      column-gap: 1rem;
      padding: 0.1rem;
    }
  }
  @media screen and (max-width: 480px) {
    &__container {
      column-count: 1;
      column-gap: 0.05rem;
      padding: 0.05rem;
    }
  }
}

.treepage {
  $self: &;
  background-color: $silver;
  height: 100vh;
  padding: 65px 10px;

  &__content {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: left;
    background-color: $white-mix;
    padding: 10px;
    border-radius: $borderRadius;
  }

  &__content-image img {
    max-height: 100%;
    min-width: 50vw;
    border-radius: $borderRadius;
  }

  .__content-image {
    width: 50%;
  }

  &__content-info {
    display: flex;
    flex-direction: column;
    justify-content: left;
    padding: 0 15px;
  }

  &__content-info-item {
    border-top: 1px solid $silver;
    font-size: medium;
    padding: 10px 0;
  }
  &__content-info-item-label {
    font-weight: 600;
  }

  @media screen and (max-width: 768px) {
    &__content {
      flex-direction: column;
      width: 100%;
    }
    &__image {
      width: 100%;
    }
  }

  @media screen and (max-width: 480px) {
    &__content {
      flex-direction: column;
      width: 100%;
    }
    &__image {
      width: 100%;
    }
  }
}

.toggles {
  background-color: $white;
  max-width: fit-content;
  border-radius: $borderRadius;
}
