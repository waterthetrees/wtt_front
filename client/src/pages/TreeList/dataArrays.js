import { topTreesCaliforniaNative } from '@/data/dist/topTreesCaliforniaNative';
import { topTreesSanFrancisco } from '@/data/dist/topTreesSanFrancisco';
import { topTreesUSFood } from '@/data/dist/topTreesUSFood';

const dataColumns = [
  {
    columnHeader: 'common',
    label: 'Common',
  },
  {
    columnHeader: 'scientific',
    label: 'Scientific',
  },
  {
    columnHeader: 'genus',
    label: 'Genus',
  },
  {
    columnHeader: 'height',
    label: 'Height',
  },
  {
    columnHeader: 'deciduousEvergreen',
    label: 'Dormancy',
  },
  {
    columnHeader: 'notes',
    label: 'Notes',
  },
];

export const dataSources = [
  {
    name: 'Native California Trees',
    data: topTreesCaliforniaNative,
    columns: dataColumns.slice(0, 5),
    url: 'https://calscape.org/loc-California/cat-Trees/ord-popular?srchcr=sc60ef7918b1949',
    thanks: '©calscape.org',
    treecare: 'https://vimeo.com/416031708#t=5m35s',
  },
  {
    name: 'Food Trees',
    data: topTreesUSFood,
    columns: dataColumns.slice(0, 3),
    url: 'https://fallingfruit.org',
    thanks: '©FallingFruit.org',
    treecare: 'https://vimeo.com/416031708#t=5m35s',
  },
  {
    name: 'San Francisco Street Trees',
    data: topTreesSanFrancisco,
    columns: dataColumns.slice(0, 6),
    url: 'https://www.fuf.net/',
    thanks: '©fuf.net',
    treecare: 'https://www.friendsoftheurbanforest.org/tree-care',
  },
];
