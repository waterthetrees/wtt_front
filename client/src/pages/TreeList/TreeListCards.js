import React from 'react';
import { Card } from '../../components/Card/Card';
import { NavLink } from 'react-router-dom';
import { ImageLoad, fixScientificForImage } from '@/pages/Tree/TreeImage';
import { Tag } from '@/components/Tag/Tag';
import { dataSources } from '@/pages/Data/dataArrays';
import './TreeList.scss';
const routeName = 'tree';

const HEADING_TEXT = `We encourage you to plant medium and larger-sized trees, as they
provide greater benefits to the city than smaller trees. Young trees
require 1.5 inches of rain or 25 gallons of water per week for the
first three years to establish their roots.`;

export default function TreeListCards({ data, selectedDataSourceIndex }) {
  return (
    <div className="treelistcards">
      <div className="treelistcards__container">
        <div className="treelistcards__header">
          <Card>
            <h3>{HEADING_TEXT}</h3>
            <div className="treecare">
              <a
                href={dataSources[selectedDataSourceIndex].treecare}
                target="_blank"
                className="treelistcards__link"
                rel="noreferrer"
              >
                Tree Care Program
              </a>
            </div>
          </Card>
        </div>

        {data.map((tree, index) => {
          const nameFormatted = tree?.common?.toLowerCase().replace(/ /g, '-');
          const urlRoute = `/${routeName}/${nameFormatted}`;
          const { scientific, common, height, deciduousEvergreen } = tree;
          const treeImagePath = fixScientificForImage(scientific);
          const tagVariant =
            deciduousEvergreen === 'deciduous' ? 'brown' : 'green';

          return (
            <NavLink
              className="treelistcards__link"
              to={{
                pathname: urlRoute,
              }}
              state={{ tree, selectedDataSourceIndex }}
              key={`${tree?.scientific}-${index}`}
            >
              <Card>
                <div className="treelistcards__image">
                  <ImageLoad
                    src={treeImagePath}
                    placeholder="placeholder.jpg"
                  />
                </div>
                <h2>{common}</h2>
                <h4>{scientific}</h4>
                <div className="treelistcards__item">{height}</div>
                {deciduousEvergreen && (
                  <Tag variant={tagVariant}>{deciduousEvergreen}</Tag>
                )}
              </Card>
            </NavLink>
          );
        })}
      </div>
    </div>
  );
}
