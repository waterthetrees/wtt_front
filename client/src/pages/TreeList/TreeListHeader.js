import React, { useState } from 'react';
import { dataSources } from './dataArrays';
import { MenuItem, FormControl, Select } from '@mui/material';
import { useIsMobile } from '../NewTree/utilities';
import { SearchBar, searchArray } from '@/components/SearchBar/SearchBar';
import { ButtonShowHide } from '@/components/Button';
import { Toggle } from '@/components/Form';
import TreeListCheckboxes from './TreeListCheckboxes';
import './TreeList.scss';

const checkboxCategories = {
  Height: {
    category: 'height',
    options: ['small', 'medium', 'large'],
  },
  Type: {
    category: 'deciduousEvergreen',
    options: ['evergreen', 'deciduous'],
  },
};

function getActiveFilters(checkboxes) {
  return Object.entries(checkboxCategories).reduce(
    (filters, [_, filterInfo]) => {
      filters[filterInfo.category] = filterOptions(
        checkboxes,
        filterInfo.options,
      );
      return filters;
    },
    {},
  );
}

function createLookup(activeFilters) {
  const lookup = {};
  for (const [category, options] of Object.entries(activeFilters)) {
    lookup[category] = options.reduce((acc, option) => {
      acc[option.toLowerCase()] = true;
      return acc;
    }, {});
  }
  return lookup;
}

export function filterData(array, checkboxes) {
  if (Object.keys(checkboxes).length === 0) return array;
  if (array.length === 0) return array;
  const activeFilters = getActiveFilters(checkboxes);
  const lookup = createLookup(activeFilters);
  return array.filter((item) => {
    return Object.entries(lookup).every(([category, options]) => {
      if (Object.keys(options).length === 0) return true;

      const itemValue = item[category]?.toLowerCase();
      if (!itemValue) return false;
      for (const option in options) {
        if (itemValue.includes(option)) return true;
      }
      return false;
    });
  });
}

function filterOptions(checkboxes, options) {
  return options.filter((option) => checkboxes[option]);
}

const dataSourceMenuItems = dataSources.map(({ name }, value) => (
  <MenuItem key={name} value={value}>
    {name}
  </MenuItem>
));

export default function TreeListHeader({
  setFilteredData,
  setSelectedDataSourceIndex,
  selectedDataSourceIndex,
  data,
  view,
  setView,
  search,
  setSearch,
  showMore,
  setShowMore,
}) {
  const isMobile = useIsMobile();
  const [checkboxes, setCheckboxes] = useState({});

  const handleDataSourceChange = (event) => {
    localStorage.setItem('setSelectedDataSourceIndex', event.target.value);
    setSelectedDataSourceIndex(event.target.value);
  };

  const handleSearch = (event) => {
    const searchValue = event.target.value;
    setSearch(searchValue);
    const searchSubset = searchArray(data, searchValue);
    const filteredDataNew = filterData(searchSubset, checkboxes);
    setFilteredData(filteredDataNew);
  };

  return (
    <div className="treelistheader">
      <div className="treelistheader__section">
        <div className="treelistheader__section-item">
          <FormControl>
            <SearchBar
              style={{
                div: {
                  width: 'max-content',
                  borderRadius: '4px',
                  fontSize: 'large',
                },
                input: { borderRadius: '4px', width: '100%' },
              }}
              search={search}
              onChange={handleSearch}
              placeholder={'Search trees'}
            />
          </FormControl>
          {isMobile && (
            <ButtonShowHide showMore={showMore} setShowMore={setShowMore} />
          )}
        </div>

        {showMore && (
          <div className="treelistheader__section-item">
            <Select
              mt={0}
              mb={0}
              MenuProps={{ disableScrollLock: true }}
              labelId="data-select-label"
              id="data-select"
              className="treelistheader__section-data-select"
              size="small"
              value={selectedDataSourceIndex}
              onChange={handleDataSourceChange}
            >
              {dataSourceMenuItems}
            </Select>
          </div>
        )}
        {showMore && (
          <div className="treelistheader__section-item-toggle">
            <Toggle view={view} setView={setView} />
          </div>
        )}
      </div>
      {showMore && (
        <TreeListCheckboxes
          setFilteredData={setFilteredData}
          data={data}
          search={search}
          checkboxes={checkboxes}
          setCheckboxes={setCheckboxes}
        />
      )}
    </div>
  );
}
