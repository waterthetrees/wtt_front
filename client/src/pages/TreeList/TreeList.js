import { clsx } from 'clsx';
import React, { useEffect, useState } from 'react';

import { ListGrid } from '@/components/ListGrid/ListGrid';
import SubHeader from '@/components/SubHeader/SubHeader';
import LicenseFooter from '@/pages/License/LicenseFooter';

import { useIsMobile } from '../NewTree/utilities';
import TreeListCards from './TreeListCards';
import TreeListHeader from './TreeListHeader';
import { dataSources } from './dataArrays';

import './TreeList.scss';

export default function TreeList() {
  const isMobile = useIsMobile();
  const [showMore, setShowMore] = useState(!isMobile);
  const [view, setView] = useState(localStorage.getItem('view') || 'card');
  const [search, setSearch] = useState('');

  const [selectedDataSourceIndex, setSelectedDataSourceIndex] = useState(
    localStorage.getItem('setSelectedDataSourceIndex') || 2,
  );
  const { url, thanks, data, columns } = dataSources[selectedDataSourceIndex];

  const [filteredData, setFilteredData] = useState(data);

  useEffect(() => {
    setFilteredData(data);
  }, [data]);

  // dynamically set padding for treelist__content on mobile
  const treeListContentStyle = clsx({
    treelist__content: showMore,
    'treelist__content-small': !showMore,
  });

  return (
    <div className="treelist">
      <div className="treelist__header">
        <SubHeader id="treelist__header">
          <TreeListHeader
            setFilteredData={setFilteredData}
            setSelectedDataSourceIndex={setSelectedDataSourceIndex}
            selectedDataSourceIndex={selectedDataSourceIndex}
            data={data}
            view={view}
            setView={setView}
            search={search}
            setSearch={setSearch}
            showMore={showMore}
            setShowMore={setShowMore}
          />
        </SubHeader>
      </div>
      <div className={treeListContentStyle}>
        {view === 'card' && (
          <TreeListCards
            data={filteredData}
            selectedDataSourceIndex={selectedDataSourceIndex}
          />
        )}
        {view === 'list' && (
          <ListGrid data={filteredData} columns={columns} listType="data" />
        )}
      </div>

      <LicenseFooter>
        <a href={url} target="_blank" rel="noreferrer">
          {thanks}
        </a>
      </LicenseFooter>
    </div>
  );
}
