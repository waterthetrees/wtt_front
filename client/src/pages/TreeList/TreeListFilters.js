import React from 'react';

import { searchArray } from '@/components/SearchBar/SearchBar';

import { filterData } from './TreeListHeader';

import './TreeList.scss';

const checkboxCategories = {
  Height: {
    category: 'height',
    options: ['small', 'medium', 'large'],
  },
  Type: {
    category: 'deciduousEvergreen',
    options: ['evergreen', 'deciduous'],
  },
};

export default function TreeListFilters({
  setFilteredData,
  data,
  search,
  setCheckboxes,
  checkboxes,
}) {
  const handleCheckboxChange = (event) => {
    const updatedCheckboxes = {
      ...checkboxes,
      [event.target.name]: event.target.checked,
    };
    // set localStorage before state for persistence before re-render
    localStorage.setItem('checkboxes', updatedCheckboxes);
    setCheckboxes(updatedCheckboxes);
    const filteredDataNew = filterData(data, updatedCheckboxes);
    const searchSubset = searchArray(filteredDataNew, search);
    setFilteredData(searchSubset);
  };

  return (
    <div className="treelistfilters">
      {Object.entries(checkboxCategories).map(([groupName, groupInfo]) => (
        <div key={groupName} className="treelistfilters__filter">
          <h3 className="treelistfilters__filter-title">{groupName}:</h3>
          <div className="treelistfilters__filter-item">
            {groupInfo.options.map((option) => (
              <label
                aria-labelledby={option}
                key={option}
                className="treelistfilters__filter-item-label"
              >
                <input
                  id={option}
                  type="checkbox"
                  name={option}
                  checked={checkboxes[option] || false}
                  onChange={handleCheckboxChange}
                  className="treelistfilters__filter-item-checkbox"
                />
                {option}
              </label>
            ))}
          </div>
        </div>
      ))}
    </div>
  );
}
