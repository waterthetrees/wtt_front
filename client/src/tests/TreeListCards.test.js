import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import TreeListCards, {
  Dormancy,
  getDormancyTagColor,
} from '@/pages/TreeList/TreeListCards';
import { dataSources } from '@/pages/TreeList/dataArrays';

/**
 * These snapshots are designed to catch _any_ change to UI components, so they
 * are by definition very brittle. If the changes are related to your changes
 * and look expected, update these snapshots with `npm test -- -u`
 */
const mockData = [
  {
    scientific: 'Quercus robur',
    common: 'English Oak',
    height: '20-30 feet',
    deciduousEvergreen: 'deciduous',
  },
  {
    scientific: 'Picea abies',
    common: 'Norway Spruce',
    height: '40-60 feet',
    deciduousEvergreen: 'evergreen',
  },
];

const mockDataSources = [
  {
    treecare: 'https://vimeo.com/416031708#t=5m35s',
  },
];

describe('TreeListCards component', () => {
  it('renders TreeListCards correctly', () => {
    const { data } = dataSources[0];
    const filteredData = data.slice(0, 2);

    const treeListCards = render(
      <Router>
        <TreeListCards data={filteredData} selectedDataSourceIndex={0} />
      </Router>,
    );
    expect(treeListCards).toMatchSnapshot();
  });

  test('renders cards correctly', () => {
    render(
      <Router>
        <TreeListCards
          data={mockData}
          selectedDataSourceIndex={0}
          dataSources={mockDataSources}
        />
      </Router>,
    );

    // Check header card
    expect(screen.getByText(/we encourage you to plant/i)).toBeInTheDocument();
    expect(screen.getByText(/tree care program/i)).toHaveAttribute(
      'href',
      'https://vimeo.com/416031708#t=5m35s',
    );

    // Check tree cards
    const links = screen.getAllByRole('link');
    expect(links).toHaveLength(3); // 2 tree cards + 1 Tree Care Program link
    expect(links[1]).toHaveAttribute('href', '/tree/quercus-robur');
    expect(links[2]).toHaveAttribute('href', '/tree/picea-abies');

    expect(screen.getByText('English Oak')).toBeInTheDocument();
    expect(screen.getByText('Norway Spruce')).toBeInTheDocument();
    expect(screen.getByText('20-30 feet')).toBeInTheDocument();
    expect(screen.getByText('40-60 feet')).toBeInTheDocument();
    expect(screen.getByText(/deciduous/i)).toBeInTheDocument();
    expect(screen.getByText(/evergreen/i)).toBeInTheDocument();
  });

  test('handles missing tree data gracefully', () => {
    const incompleteData = [
      {
        scientific: 'Quercus robur',
      },
    ];

    render(
      <Router>
        <TreeListCards
          data={incompleteData}
          selectedDataSourceIndex={0}
          dataSources={mockDataSources}
        />
      </Router>,
    );

    expect(screen.getByText(/we encourage you to plant/i)).toBeInTheDocument();
    expect(screen.getByText(/tree care program/i)).toHaveAttribute(
      'href',
      'https://vimeo.com/416031708#t=5m35s',
    );

    const links = screen.getAllByRole('link');
    expect(links).toHaveLength(2); // 1 tree card + 1 Tree Care Program link
    expect(links[1]).toHaveAttribute('href', '/tree/quercus-robur');
  });
});

describe('Dormancy component', () => {
  test('renders nothing if no valid string is provided', () => {
    render(<Dormancy deciduousEvergreen={null} />);
    expect(screen.queryByRole('tag')).not.toBeInTheDocument();

    render(<Dormancy deciduousEvergreen={123} />);
    expect(screen.queryByRole('tag')).not.toBeInTheDocument();

    render(<Dormancy />);
    expect(screen.queryByRole('tag')).not.toBeInTheDocument();
  });

  test('renders single dormancy tag', () => {
    render(<Dormancy deciduousEvergreen="Evergreen" />);
    const tag = screen.getByText('evergreen');
    expect(tag).toBeInTheDocument();
  });

  test('renders multiple dormancy tags', () => {
    render(<Dormancy deciduousEvergreen="Deciduous,Winter Deciduous" />);
    const deciduousTag = screen.getByText('deciduous');
    const evergreenTag = screen.getByText('winter-deciduous');
    expect(deciduousTag).toBeInTheDocument();
    expect(evergreenTag).toBeInTheDocument();
  });
});

describe('getDormancyTagColor', () => {
  it('returns green for evergreen', () => {
    expect(getDormancyTagColor('evergreen')).toBe('green');
  });

  it('returns brown for deciduous', () => {
    expect(getDormancyTagColor('deciduous')).toBe('brown');
  });

  it('returns blue for winter-deciduous', () => {
    expect(getDormancyTagColor('winter-deciduous')).toBe('blue');
  });

  it('returns brown for summer-deciduous', () => {
    expect(getDormancyTagColor('summer-deciduous')).toBe('brown');
  });

  it('returns brown for summer-semi-deciduous', () => {
    expect(getDormancyTagColor('summer-semi-deciduous')).toBe('brown');
  });

  it('returns black for unknown dormancy', () => {
    expect(getDormancyTagColor('unknown')).toBe('black');
  });

  it('returns black for no dormancy provided', () => {
    expect(getDormancyTagColor()).toBe('black');
  });
});
