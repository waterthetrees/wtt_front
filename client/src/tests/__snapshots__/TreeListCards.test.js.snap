// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeListCards component renders TreeListCards correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="treelistcards"
      >
        <div
          class="treelistcards__container"
        >
          <div
            class="treelistcards__header"
          >
            <div
              class="card"
              data-testid="card"
            >
              <div
                class="treelistcards__info"
              >
                <h3>
                  We encourage you to plant medium and larger-sized trees, as they
provide greater benefits to the city than smaller trees. Young trees
require 1.5 inches of rain or 25 gallons of water per week for the
first three years to establish their roots.
                </h3>
                <div
                  class="treecare"
                >
                  <a
                    class="treelistcards__link"
                    href="https://vimeo.com/416031708#t=5m35s"
                    rel="noreferrer"
                    target="_blank"
                  >
                    Tree Care Program
                  </a>
                </div>
              </div>
            </div>
          </div>
          <a
            class="treelistcards__link"
            href="/tree/sequoia-sempervirens-'adpressa'"
          >
            <div
              class="card"
              data-testid="card"
            >
              <div
                class="treelistcards__image"
              >
                <img
                  alt="Sequoia sempervirens 'Adpressa'"
                  loading="lazy"
                  placeholder="./../assets/images/treelist/placeholder.jpg"
                  style="opacity: 0.1; transition: opacity .30s linear; width: 100%;"
                />
              </div>
              <div
                class="treelistcards__info"
              >
                <h2>
                  Adpressa Redwood
                </h2>
                <h4
                  class="scientific"
                >
                  Sequoia sempervirens 'Adpressa'
                </h4>
                <div
                  class="treelistcards__item"
                >
                  100 - 200 ft (30.5 - 61 m)
                </div>
                <div
                  class="tag"
                >
                  <div
                    class="tag__pill tag__green"
                  >
                    evergreen
                  </div>
                </div>
              </div>
            </div>
          </a>
          <a
            class="treelistcards__link"
            href="/tree/callitropsis-nootkatensis"
          >
            <div
              class="card"
              data-testid="card"
            >
              <div
                class="treelistcards__image"
              >
                <img
                  alt="Callitropsis nootkatensis"
                  loading="lazy"
                  placeholder="./../assets/images/treelist/placeholder.jpg"
                  style="opacity: 0.1; transition: opacity .30s linear; width: 100%;"
                />
              </div>
              <div
                class="treelistcards__info"
              >
                <h2>
                  Alaska Cedar
                </h2>
                <h4
                  class="scientific"
                >
                  Callitropsis nootkatensis
                </h4>
                <div
                  class="treelistcards__item"
                >
                  65.6 - 98.4 ft(20 - 30 m)
                </div>
                <div
                  class="tag"
                >
                  <div
                    class="tag__pill tag__green"
                  >
                    evergreen
                  </div>
                </div>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="treelistcards"
    >
      <div
        class="treelistcards__container"
      >
        <div
          class="treelistcards__header"
        >
          <div
            class="card"
            data-testid="card"
          >
            <div
              class="treelistcards__info"
            >
              <h3>
                We encourage you to plant medium and larger-sized trees, as they
provide greater benefits to the city than smaller trees. Young trees
require 1.5 inches of rain or 25 gallons of water per week for the
first three years to establish their roots.
              </h3>
              <div
                class="treecare"
              >
                <a
                  class="treelistcards__link"
                  href="https://vimeo.com/416031708#t=5m35s"
                  rel="noreferrer"
                  target="_blank"
                >
                  Tree Care Program
                </a>
              </div>
            </div>
          </div>
        </div>
        <a
          class="treelistcards__link"
          href="/tree/sequoia-sempervirens-'adpressa'"
        >
          <div
            class="card"
            data-testid="card"
          >
            <div
              class="treelistcards__image"
            >
              <img
                alt="Sequoia sempervirens 'Adpressa'"
                loading="lazy"
                placeholder="./../assets/images/treelist/placeholder.jpg"
                style="opacity: 0.1; transition: opacity .30s linear; width: 100%;"
              />
            </div>
            <div
              class="treelistcards__info"
            >
              <h2>
                Adpressa Redwood
              </h2>
              <h4
                class="scientific"
              >
                Sequoia sempervirens 'Adpressa'
              </h4>
              <div
                class="treelistcards__item"
              >
                100 - 200 ft (30.5 - 61 m)
              </div>
              <div
                class="tag"
              >
                <div
                  class="tag__pill tag__green"
                >
                  evergreen
                </div>
              </div>
            </div>
          </div>
        </a>
        <a
          class="treelistcards__link"
          href="/tree/callitropsis-nootkatensis"
        >
          <div
            class="card"
            data-testid="card"
          >
            <div
              class="treelistcards__image"
            >
              <img
                alt="Callitropsis nootkatensis"
                loading="lazy"
                placeholder="./../assets/images/treelist/placeholder.jpg"
                style="opacity: 0.1; transition: opacity .30s linear; width: 100%;"
              />
            </div>
            <div
              class="treelistcards__info"
            >
              <h2>
                Alaska Cedar
              </h2>
              <h4
                class="scientific"
              >
                Callitropsis nootkatensis
              </h4>
              <div
                class="treelistcards__item"
              >
                65.6 - 98.4 ft(20 - 30 m)
              </div>
              <div
                class="tag"
              >
                <div
                  class="tag__pill tag__green"
                >
                  evergreen
                </div>
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
